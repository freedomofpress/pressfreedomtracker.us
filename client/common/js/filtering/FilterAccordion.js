import React from 'react'
import classNames from 'classnames'
import { CollapseIcon, ExpandIcon } from '~/filtering/Icons'
import FilterSets from '~/filtering/FilterSets'
import NewFilterSet from '~/filtering/NewFilterSet'


function FilterAccordion({
	category,
	selectedAccordions,
	handleAccordionSelection,
	handleFilterChange,
	filterValues,
	noCategoryFiltering,
	choices,
}) {
	if (!category.enabled) {
		return null
	}
	let FilterSet
	if(category.id === -1) {
		// General is a special category with id -1 that should be rendered
		FilterSet = typeof FilterSets[category.title] === 'function' ? FilterSets[category.title] : null
	} else if (!category.related_fields || !(category.related_fields.length > 0)) {
		// Don't bother rendering an accordion with no filters
		return null
	}


	const isActive = selectedAccordions.includes(category.id)
	const collapsible = !noCategoryFiltering

	let renderedFilterset
	if(isActive && category.related_fields) {
		renderedFilterset = (<NewFilterSet
			title={category.title}
			fields={category.related_fields}
			handleFilterChange={handleFilterChange}
			filterValues={filterValues}
			choices={choices}
			key={category.id}
		/>)
	} else if(isActive && FilterSet) {
		renderedFilterset = (
			<FilterSet
				handleFilterChange={handleFilterChange}
				filterValues={filterValues}
			/>
		)
	} else {
		renderedFilterset = null
	}

	return (
		<li
			className={classNames(
				'filters__accordion-category',
				{ 'filters__accordion-category--no-divider': noCategoryFiltering }
			)}
		>
			{collapsible && (
				<button
					className={classNames(
						'filters__accordion',
						{ 'filters__accordion--active': isActive }
					)}
					onClick={handleAccordionSelection.bind(null, category.id)}
					type="button"
				>
					{isActive ? <CollapseIcon /> : <ExpandIcon />}
					{category.title}
				</button>
			)}

			{renderedFilterset}
		</li>
	)
}


export default FilterAccordion
