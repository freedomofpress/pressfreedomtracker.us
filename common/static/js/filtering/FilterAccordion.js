import React from 'react'
import classNames from 'classnames'
import { CollapseIcon, ExpandIcon } from '~/filtering/Icons'
import FilterSets from '~/filtering/FilterSets'


function FilterAccordion({
	category,
	selectedAccordions,
	handleAccordionSelection,
	handleFilterChange,
	filterValues,
	noCategoryFiltering,
	choices,
}) {
	if (!category.enabled) {
		return null
	}

	const FilterSet = typeof FilterSets[category.title] === 'function' ? FilterSets[category.title] : null
	if (!FilterSet) {
		// Don't bother rendering an accordion with no filters
		return null
	}

	const isActive = selectedAccordions.includes(category.id)
	const collapsible = !noCategoryFiltering

	return (
		<li
			className={classNames(
				'filters__accordion-category',
				{ 'filters__accordion-category--no-divider': noCategoryFiltering }
			)}
		>
			{collapsible && (
				<button
					className={classNames(
						'filters__accordion',
						{ 'filters__accordion--active': isActive }
					)}
					onClick={handleAccordionSelection.bind(null, category.id)}
				>
					{isActive ? <CollapseIcon /> : <ExpandIcon />}
					{category.title}
				</button>
			)}

			{isActive && (
				<FilterSet
					handleFilterChange={handleFilterChange}
					filterValues={filterValues}
					choices={choices}
				/>
			)}
		</li>
	)
}


export default FilterAccordion
