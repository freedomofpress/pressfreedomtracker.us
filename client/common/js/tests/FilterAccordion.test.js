import React from 'react';
import { shallow, mount } from 'enzyme';
import FilterAccordion from '~/filtering/FilterAccordion';
import {
    CATEGORIES_FILTER,
	FILTER_VALUES,
} from '~/tests/factories/CategoriesFilterFactory';

describe('FilterAccordion', () => {
    let categories;
	let filterValues;

    beforeEach(() => {
        categories = CATEGORIES_FILTER;
        filterValues = FILTER_VALUES;
    });

    afterEach(() => {
        categories = null;
        filterValues = null;
    });

    it('should render without error in "debug" mode', () => {
        const component = shallow(
            <FilterAccordion
                category={categories[0]}
                collapsible={true}
                handleFilterChange={jest.fn()}
                filterValues={filterValues}
                startExpanded={true}
            />
        );

        expect(component.debug()).toMatchSnapshot();
	});

	it('should render FilterAccordion properly with collapsible', () => {
		const component = mount(
            <FilterAccordion
                category={categories[0]}
                collapsible={true}
                handleFilterChange={jest.fn()}
                filterValues={filterValues}
                startExpanded={false}
            />
        );
        let filterAccordion = component.find('.filters__accordion');
        expect(filterAccordion.length).toEqual(1);
        expect(filterAccordion.find('.expand__icon').length).toEqual(1);

        const nonCollapsibleComponent = mount(
            <FilterAccordion
                category={categories[0]}
                collapsible={false}
                handleFilterChange={jest.fn()}
                filterValues={filterValues}
                startExpanded={false}
            />
        );
        filterAccordion = nonCollapsibleComponent.find('.filters__accordion');
        expect(filterAccordion.length).toEqual(0);

        const collapsibleExpandedComponent = mount(
            <FilterAccordion
                category={categories[0]}
                collapsible={true}
                handleFilterChange={jest.fn()}
                filterValues={filterValues}
                startExpanded={true}
            />
        );
        filterAccordion = collapsibleExpandedComponent.find('.filters__accordion');
        expect(filterAccordion.find('.collapse__icon').length).toEqual(1);
    });

    it('should render FilterAccordion properly with expanded', () => {
		const component = mount(
            <FilterAccordion
                category={categories[0]}
                collapsible={false}
                handleFilterChange={jest.fn()}
                filterValues={filterValues}
                startExpanded={true}
            />
        );
        const filterAccordion = component.find('.filters__set');
        expect(filterAccordion.length).toEqual(1);
    });
});