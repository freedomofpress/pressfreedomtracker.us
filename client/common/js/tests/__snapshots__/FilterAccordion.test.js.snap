// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterAccordion should render without error in "debug" mode 1`] = `
"<li className=\\"filters__accordion-category\\">
  <button className=\\"filters__accordion filters__accordion--active\\" onClick={[Function: bound toggleAccordion]} type=\\"button\\">
    <CollapseIcon />
    General
  </button>
  <FilterSet filters={{...}} handleFilterChange={[Function: mockConstructor]} filterValues={{...}} />
</li>"
`;
